File: fractals/cpp/mandelbrot.cpp
#include <opencv2/opencv.hpp>
#include <complex>

int mandelbrot(double x0, double y0, int max_iter) {
    std::complex<double> z(0, 0);
    std::complex<double> c(x0, y0);
    int iter = 0;
    while (std::abs(z) <= 2.0 && iter < max_iter) {
        z = z * z + c;
        ++iter;
    }
    return iter;
}

int main() {
    const int width = 800;
    const int height = 600;
    const int max_iter = 1000;

    cv::Mat image(height, width, CV_8UC3);

    for (int py = 0; py < height; ++py) {
        for (int px = 0; px < width; ++px) {
            double x0 = (px - width/2.0) * 4.0 / width;
            double y0 = (py - height/2.0) * 4.0 / width;

            int iter = mandelbrot(x0, y0, max_iter);

            int color = static_cast<int>(255.0 * iter / max_iter);
            image.at<cv::Vec3b>(py, px) = cv::Vec3b(0, color, color);  // azul/vermelho
        }
    }

    cv::imwrite("mandelbrot.png", image);
    return 0;
}

File: fractals/julia/mandelbrot.jl
using Plots

function mandelbrot(c; max_iter=100)
    z = 0 + 0im
    for n = 1:max_iter
        z = z^2 + c
        if abs(z) > 2
            return n
        end
    end
    return max_iter
end

function generate_fractal(xmin, xmax, ymin, ymax, res, max_iter)
    x = range(xmin, xmax; length=res)
    y = range(ymin, ymax; length=res)
    image = zeros(Int, res, res)
    for i in 1:res, j in 1:res
        c = complex(x[j], y[i])
        image[i, j] = mandelbrot(c; max_iter=max_iter)
    end
    return image
end

img = generate_fractal(-2.0, 1.0, -1.5, 1.5, 500, 100)
heatmap(img, color=:viridis, legend=false)